cpp/comms/std/src/cuML_std_comms_impl.cpp

```diff
@@ -0,0 +1,408 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "cuML_std_comms_impl.hpp"
+
+#include <nccl.h>
+
+#ifdef WITH_UCX
+constexpr bool UCX_ENABLED = true;
+#else
+constexpr bool UCX_ENABLED = false;
+#endif
+
+#ifdef WITH_UCX
+#include <pthread.h>
+#include <ucp/api/ucp.h>
+#include <ucp/api/ucp_def.h>
+#include "ucp_helper.h"
+#endif
+
+#include <cstdio>
+#include <exception>
+#include <memory>
+
+#include <common/cumlHandle.hpp>
+#include <cuML_comms.hpp>
+
+#include <cuda_runtime.h>
+
+#include <utils.h>
+
+#define NCCL_CHECK(call)                                                       \
+  do {                                                                         \
+    ncclResult_t status = call;                                                \
+    ASSERT(ncclSuccess == status, "ERROR: NCCL call='%s'. Reason:%s\n", #call, \
+           ncclGetErrorString(status));                                        \
+  } while (0)
+
+//@todo adapt logging infrastructure for NCCL_CHECK_NO_THROW once available:
+//https://github.com/rapidsai/cuml/issues/100
+#define NCCL_CHECK_NO_THROW(call)                                              \
+  do {                                                                         \
+    ncclResult_t status = call;                                                \
+    if (ncclSuccess != status) {                                               \
+      std::fprintf(stderr,                                                     \
+                   "ERROR: NCCL call='%s' at file=%s line=%d failed with %s ", \
+                   #call, __FILE__, __LINE__, ncclGetErrorString(status));     \
+    }                                                                          \
+  } while (0)
+
+namespace ML {
+
+namespace {
+
+size_t getDatatypeSize(const cumlStdCommunicator_impl::datatype_t datatype) {
+  switch (datatype) {
+    case MLCommon::cumlCommunicator::CHAR:
+      return sizeof(char);
+    case MLCommon::cumlCommunicator::UINT8:
+      return sizeof(unsigned char);
+    case MLCommon::cumlCommunicator::INT:
+      return sizeof(int);
+    case MLCommon::cumlCommunicator::UINT:
+      return sizeof(unsigned int);
+    case MLCommon::cumlCommunicator::INT64:
+      return sizeof(long long int);
+    case MLCommon::cumlCommunicator::UINT64:
+      return sizeof(unsigned long long int);
```

Comment:
Same as above - `int64_t` and `uint64_t` ?

Reply:
@jirikraus, thoughts?

Reply:
Agree the `int*_t` types defined in `cstdint` are a better choice.